services/common/rs_server_common/authentication/authentication_to_external.py

```diff
@@ -0,0 +1,458 @@
+            status_code=HTTP_500_INTERNAL_SERVER_ERROR,
+            detail=f"Request to token endpoint failed: {str(e)}",
+        ) from e
+
+    token = response.json()
+    # TODO: Is it worthy to validate it?
+    # validate_token_format(token.get("access_token", ""))
+    if ACCESS_TK_KEY_IN_RESPONSE not in token:
+        logger.error(
+            f"The token field was not found in the response from the station {external_auth_config.station_id}. ",
+        )
+        raise HTTPException(
+            status_code=HTTP_404_NOT_FOUND,
+            detail=f"The token field was not found in the response from the station {external_auth_config.station_id}.",
+        )
+    logger.info(f"Access token retrieved from the station url: {external_auth_config.token_url} ")
+    return token.get("access_token")
+
+
+def prepare_headers(external_auth_config: ExternalAuthenticationConfig) -> Dict[str, str]:
+    """Prepare HTTP headers for token requests.
+
+    Args:
+        external_auth_config (ExternalAuthenticationConfig): Configuration object containing authentication details.
+
+    Returns:
+        Dict[str, str]: Dictionary containing the prepared headers.
+    """
+    headers = {"Content-Type": HEADER_CONTENT_TYPE}
+    # Add Authorization header if it exists
+    if external_auth_config.authorization:
+        headers["Authorization"] = external_auth_config.authorization
+    return headers
+
+
+def prepare_data(external_auth_config: ExternalAuthenticationConfig) -> Dict[str, str]:
+    """Prepare data for token requests based on authentication configuration.
+
+    Args:
+        external_auth_config (ExternalAuthenticationConfig): Configuration object containing authentication details.
+
+    Returns:
+        Dict[str, str]: Dictionary containing the prepared data for the request.
+    """
+    data_to_send = {
+        "client_id": external_auth_config.client_id,
+        "client_secret": external_auth_config.client_secret,
+        "grant_type": external_auth_config.grant_type,
+        "username": external_auth_config.username,
+        "password": external_auth_config.password,
+    }
+    if external_auth_config.scope:
+        data_to_send["scope"] = external_auth_config.scope
+    return data_to_send
+
+
+def validate_token_format(token: str) -> None:
+    """Validate the format of a given token.
+
+    Args:
+        token (str): The token string to be validated.
+
+    Raises:
+        HTTPException: If the token format does not match the expected pattern.
+    """
+    # Check if the token matches the expected format using a regular expression
+    if not re.match(r"^[A-Za-z0-9\-_\.]+$", token):
+        # Raise an HTTP exception if the token format is invalid
+        logger.error("Invalid token format received from the station.")
+        raise HTTPException(status_code=HTTP_400_BAD_REQUEST, detail="Invalid token format received from the station.")
+
+
+def load_external_auth_config_by_station_service(
```

Comment:
Feel likes this 2 function are big-time duplicate code, can/should be merged into one later.

```python
def load_external_auth_config(
    lookup_value: str,
    lookup_type: str = "station_service",
    path: str | None = None,
) -> Optional[ExternalAuthenticationConfig]:
    """
    Load the external authentication configuration from a YAML file.
    
    Args:
        lookup_value (str): The value to look up (station_id/service or domain).
        lookup_type (str): The type of lookup to perform. Can be 'station_service' or 'domain'.
        path (str | None): Optional path to the configuration file.

    Returns:
        Optional[ExternalAuthenticationConfig]: An object representing the external authentication configuration,
        or None if no matching configuration is found or if an error occurs.

    Raises:
        HTTPException: For any errors during the loading process.
    """
    config_file = f"{path.rstrip('/')}/{CONFIG_FILENAME}" if path else DEFAULT_CONFIG_PATH_AUTH_TO_EXTERNAL

    try:
        with open(config_file, encoding="utf-8") as f:
            config_yaml = yaml.safe_load(f)
            logger.info(f"Loaded configuration YAML: {config_yaml}")
    except (FileNotFoundError, yaml.YAMLError) as e:
        logger.error(f"Error loading configuration: {e}")
        raise HTTPException(
            status_code=500,  # HTTP_500_INTERNAL_SERVER_ERROR
            detail=f"Error loading configuration. {e}",
        ) from e
    except Exception as e:
        logger.exception(f"An unexpected error occurred: {e}")
        raise HTTPException(
            status_code=500,  # HTTP_500_INTERNAL_SERVER_ERROR
            detail=f"An unexpected error occurred. {e}",
        ) from e

    # Retrieve configuration based on lookup type
    if lookup_type == "station_service":
        station_id, service = lookup_value.split('/')
        station_dict = config_yaml.get("external_data_sources", {}).get(station_id, {})
        service_dict = station_dict.get("service", {})

        if service_dict.get("name") != service:
            logger.warning(f"No matching service found for station_id: {station_id} and service: {service}")
            return None

        return create_external_auth_config(station_id, station_dict, service_dict)

    elif lookup_type == "domain":
        for station_id, station_dict in config_yaml.get("external_data_sources", {}).items():
            if station_dict.get("domain") == lookup_value:
                return create_external_auth_config(station_id, station_dict, station_dict.get("service", {}))

    # No matching configuration found
    logger.warning(f"No matching configuration found for {lookup_type}: {lookup_value}")
    return None

```